src/MediaUploader.php

```diff
@@ -438,12 +453,51 @@ public function upload()
 
         $this->verifyDestination($model);
 
-        $this->filesystem->disk($model->disk)->put($model->getDiskPath(), $this->source->contents());
+        $source = $this->callBefore('before_upload', $model, SourceAdapterInterface::class);
+
+        if ($source !== null) {
+            $model->size = $source->size();
+            $model->mime_type = $source->mimeType();
+            $model->extension = $source->extension();
+            $model->aggregate_type = $this->inferAggregateType($model->mime_type, $model->extension);
+        } else {
+            $source = $this->source;
+        }
+
+        $alteredModel = $this->callBefore('before_save', $model, Model::class);
+
+        if ($alteredModel !== null) {
+            $model = $alteredModel;
+        }
+
+        $this->filesystem->disk($model->disk)->put($model->getDiskPath(), $source->contents());
         $model->save();
 
         return $model;
     }
 
+    /**
+     * Set the before upload callback
+     * @param callable $callable
+     * @return static
+     */
+    public function beforeUpload($callable)
+    {
+        $this->before_upload = $callable;
+        return $this;
+    }
+
+    /**
+     * Set the before save callback
+     * @param callable $callable
+     * @return static
+     */
+    public function beforeSave($callable)
```

Comment:
same

Reply:
Never sure about what can and cannot be typehint-ed in older php versions. Done.